bindings/rust/s2n-tls-tokio/src/lib.rs

```diff
@@ -404,19 +363,26 @@ where
     fn poll_shutdown(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<io::Result<()>> {
         ready!(self.as_mut().poll_blinding(ctx))?;
 
-        let status = ready!(self.as_mut().with_io(ctx, |mut context| {
-            context.conn.as_mut().poll_shutdown().map(|r| r.map(|_| ()))
-        }));
+        // s2n_shutdown should not be called again if it errors
+        if self.shutdown_error.is_none() {
+            let result = ready!(self.as_mut().with_io(ctx, |mut context| {
+                context.conn.as_mut().poll_shutdown().map(|r| r.map(|_| ()))
+            }));
+            if let Err(error) = result {
+                self.shutdown_error = Some(error);
+                // s2n_shutdown reading might have triggered blinding again
+                ready!(self.as_mut().poll_blinding(ctx))?;
+            }
+        };
 
-        if let Err(e) = status {
-            // In case of an error shutting down, make sure you wait for
-            // the blinding timeout.
-            self.as_mut().set_blinding_timer(Err(e))?;
-            ready!(self.as_mut().poll_blinding(ctx))?;
-            unreachable!("should have returned the error we just put in!");
-        }
+        let tcp_result = ready!(Pin::new(&mut self.as_mut().stream).poll_shutdown(ctx));
 
-        Pin::new(&mut self.as_mut().stream).poll_shutdown(ctx)
+        let result = if let Some(error) = self.shutdown_error.take() {
```

Comment:
Trying to decide if it's best to call `take` or `cloned`. If we take the error then this function can't be called again without getting into a weird state. It is kind of nice to ["fuse"](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.fuse) poll functions so they return the same Ready value, although it's not strictly required by the contract:

https://doc.rust-lang.org/beta/std/future/trait.Future.html#return-value

> Once a future has finished, clients should not poll it again.

https://doc.rust-lang.org/beta/std/future/trait.Future.html#panics

> Once a future has completed (returned Ready from poll), calling its poll method again may panic, block forever, or cause other kinds of problems; the Future trait places no requirements on the effects of such a call. However, as the poll method is not marked unsafe, Rust’s usual rules apply: calls must never cause undefined behavior (memory corruption, incorrect use of unsafe functions, or the like), regardless of the future’s state.

Reply:
I can't actually clone the error, but I can swap in a new error.